moveit-core: Fixes for Windows, and CI for macOS and Windows

wolfv opened the pull request:
### Description

Hi Moveit team!

This PR has the fixes we had to apply in the RoboStack project for noetic (http://github.com/robostack/ros-noetic). In the RoboStack project we build conda packages for ROS on all major OS's: macOS, Windows and Linux.

This PR mostly applies patches by @seanyen from Microsoft. 

Using the conda packages from RoboStack also makes it easy to set up a cross-platform CI which we've done in this PR as well.

### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit/blob/master/MIGRATION.md) notes
- [ ] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/doc/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

wolfv commented:
pinging @simonschmeisser and @Tobias-Fischer 

ref: https://github.com/ros-planning/moveit/issues/2548

simonschmeisser commented on moveit_core/robot_model/include/moveit/robot_model/fixed_joint_model.h:
```diff
@@ -52,7 +52,7 @@ class FixedJointModel : public JointModel
                                   const Bounds& other_bounds) const override;
   void getVariableRandomPositionsNearBy(random_numbers::RandomNumberGenerator& rng, double* values,
-                                        const Bounds& other_bounds, const double* near,
+                                        const Bounds& other_bounds, const double* _near,
```
Could you split this commit into smaller more descriptive ones and describe why strange prefixes are needed here?

wolfv replied:
Apparently `far` and `near` used to be keywords: https://stackoverflow.com/questions/8948493/visual-studio-doesnt-allow-me-to-use-certain-variable-names

We can also add `#undef near` to the beginning of this file. Do you prefer that?

simonschmeisser replied:
oh wow, they never cease to impress me ... but who includes `windows.h` here?

wolfv replied:
maybe rostime or some other dependency? Maybe this was also done by replacing all occurrences of near globally.
We can try to keep the patches more minimal for upstream ... now that we have a working CI for Windows.

simonschmeisser replied:
I'd prefer a
```
#if  msvc
#undef near
#endif
```
here or as close to the pollutant yes

v4hn replied:
What's the big deal with changing a variable name?
`near` is a bad name for the value anyway and I'd prefer to just change it to `seed` or `mean`.

We should not add MS specific code unless that is really needed.

wolfv replied:
completely up to you, happy to make either change.

tylerjw replied:
I would prefer renaming this variable to something else instead of the underscore prefix.

wolfv replied:
you guys just have to tell me which name to choose?

Should I go with `seed`?

v4hn replied:
Yes, just use `seed` please.

wolfv replied:
rename to seed done.

simonschmeisser commented on moveit_core/constraint_samplers/test/pr2_arm_ik.cpp:
```diff
@@ -785,5 +785,5 @@ bool PR2ArmIK::checkJointLimits(const double& joint_value, const int& joint_num)
     jv = angles::normalize_angle(joint_value * angle_multipliers_[joint_num]);
 
-  return not(jv < min_angles_[joint_num] || jv > max_angles_[joint_num]);
+  return !(jv < min_angles_[joint_num] || jv > max_angles_[joint_num]);
```
oh the classic inability of MS to properly implement C++ ...

v4hn replied:
I'm not a fan of mixing the operator types anyway (why `not`, but not `or`), so we might as well fix it.

simonschmeisser replied:
sure, I just think C++ would look half as ugly if `not`, `and`, `or` could be used in instead of "random" special characters ... but since "Basic for MS Dos" back in 1843 screwed that up now we cannot ever use that ...

simonschmeisser replied:
It seems you just have to disable msvc specific extensions which sounds like a good idea generally: https://docs.microsoft.com/en-us/cpp/build/reference/permissive-standards-conformance?view=msvc-160&viewFallbackFrom=vs-2019

v4hn replied:
I would have expected [this cmake option](https://github.com/ros-planning/moveit/blob/master/moveit_core/CMakeLists.txt#L8), but apparently msvc's non-permissive mode is something different than standard extensions...

simonschmeisser replied:
https://gitlab.kitware.com/cmake/cmake/-/issues/17068 talks about that but is still open, it also says this is the default now in VS2017 projects

wolfv replied:
I think you should debate that with the catkin folks or whoever sets global variables like that.

Since this is the only "not" in the codebase (and also the only one in all of ROS as far as I can see) I think the best course of action is to change it to !

simonschmeisser replied:
Sure, no real objections for this specific case here, my remarks were more of a general nature.

simonschmeisser commented on moveit_core/CMakeLists.txt:
```diff
@@ -46,16 +42,8 @@ else()
-  list(APPEND LIBFCL_LIBRARIES ${_lib_${_lib}})
-endforeach()
-
+find_package(fcl REQUIRED)
```
the version of fcl in Ubuntu bionic did not provide a working CMakeConfig, so this is still needed for backwards compatibility I think.

wolfv replied:
yep, I'll fix that today!

simonschmeisser reviewed (changes requested):
Please give some more info why you need to change what things. If you separate out trivial commits like the missing dependencies those could maybe also be merged earlier separately

wolfv commented:
@simonschmeisser thanks for having a look! I think there are not so many really "deep" changes in this PR. I will fix the CMake so that it works with both the Conda-provided packages and Ubuntu / ROS packages. 

Besides the `near` stuff, the biggest change is adding some "export" flags to properly export some symbols / variables in Windows DLLs. But that's also a fairly mechanical change.

wolfv commented:
By the way, the working build / test for moveit core is executed here: https://github.com/RoboStack/moveit/actions/runs/751121040

v4hn commented on .github/workflows/win.yml:
```diff
@@ -0,0 +1,114 @@
+        ninja
+        ninja run_tests
+
+    - name: Build RViz on Windows
```
rviz?

wolfv replied:
yeah, classic copy paste error :)

v4hn commented on moveit_core/collision_detection/include/moveit/collision_detection/allvalid/collision_detector_allocator_allvalid.h:
```diff
@@ -39,13 +39,23 @@
 {
 public:
-  static const std::string NAME;  // defined in collision_env_allvalid.cpp
+  MOVEIT_COLLISION_DETECTION_DECL static const std::string NAME;  // defined in collision_env_allvalid.cpp
```
how about changing it to `static constexpr char[] NAME { ... };` instead? does that still require the helper?

wolfv replied:
yes, I think it will still need `__dllexport` in front of the static variable for it to be properly exported into the DLL (and usable from other libraries).

v4hn replied:
ttbomk the NAME strings *should not* be used outside the class, they are just used to name `ROS_*` log methods in the class.
So I'm wondering why these are a public interface anyway?

`static constexpr char[] NAME` in the corresponding sources seems like the better approach here (and wouldn't need these cumbersome exports).

v4hn replied:
Note: If we really need the DECL macros, we should provide the helpers in moveit_core/macros and not require pre-processor logic in every single module.

tylerjw replied:
I think what you caught here was us having that NAME variable defined in the class when (imo) it should just be constexpr in the source file as it is not part of the external interface and is only for log namespacing as @v4hn said.  I can try to fix these for you and make a small PR against your branch for that.

wolfv replied:
that would be great. I just fixed pre-commit and tried to make our CMake changes backward compatible. let's see what the combined CI's say.

tylerjw replied:
I finally got around to looking at this and this `NAME` variable is not what we thought it was @v4hn.  It is not used in ROS_*_LOG messages and is instead part of the public interface of this class.  Each of the `collision_detection::CollisionDetectorAllocator*` classes have a NAME field.  These are used by the getName field and is part of the public interface.  This seems odd to me but it isn't what we thought it was.

I'm not sure what the right approach is.  I know for moveit2 the approach we saw taken was to define these macros in a separate moveit_common package that all the packages could depend on and get the export symbol header from there.

Have you tried the cmake option: `set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)` as that would make it simpler if it worked.    It looks like [from the documentation](https://cmake.org/cmake/help/latest/prop_tgt/WINDOWS_EXPORT_ALL_SYMBOLS.html) to use this you'd need cmake 3.4 or newer on windows.  Would that be a possible option?

Tobias-Fischer replied:
See https://github.com/ros-planning/moveit2/pull/286 for a discussion around `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS`

v4hn replied:
@tylerjw In this case I would argue we should refactor the API to a simple virtual getter method instead.
Exposing a static string as API is awkward at least and brings us more trouble than necessary in these efforts.

tylerjw replied:
I agree that the static string is awkward and should just be refactored away.

Tobias-Fischer replied:
This can be resolved.

tylerjw commented:
The [pre-commit format](https://github.com/ros-planning/moveit/pull/2604/checks?check_run_id=2350493388) test is failing on this.  Fortunately that is really easy to fix.  You just need to install pre-commit (using python3 pip) and clang-format-10 (using apt or however you'd like to install it) locally and run `pre-commit run -a` in the repo and it should fix all formatting errors.  For errors it can't fix it'll just output the result of the various linters it runs (most are auto formatting).

tylerjw commented on .github/workflows/win.yml:
```diff
@@ -0,0 +1,114 @@
+      uses: mamba-org/provision-with-micromamba@main
+      with:
+        environment-file: .github/ci_env.yml
+    # - name: Install System Dependencies
+    #   if: runner.os == 'Linux'
+    #   run: |
+    #     sudo apt-get install libglew-dev mesa-utils xvfb libgl1-mesa-glx libgl1-mesa-dev libglu1-mesa libglu1-mesa-dev \
+    #                          libselinux1 libxdamage1 libxfixes3 libxxf86vm1
```
Should this be commented out?  If so maybe it can just be removed?

tylerjw commented on .github/workflows/win.yml:
```diff
@@ -0,0 +1,114 @@
+      if: runner.os == 'Linux' || runner.os == 'macOS'
+      shell: bash -l -eo pipefail {0}
+      run: |
+        # if [[ `uname -s` == "Linux" ]]; then
+        #   CDT="-cos6-x86_64"
+        #   micromamba install -y mesa-libgl-devel$CDT mesa-dri-drivers$CDT \
+        #                         libselinux$CDT libxdamage$CDT libxxf86vm$CDT \
+        #                         libxext$CDT libxfixes$CDT -c conda-forge
+        # fi
```
Again, more commented out code.

tylerjw commented on .github/workflows/win.yml:
```diff
@@ -0,0 +1,114 @@
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-latest, macos-latest, windows-latest]
```
Since this runs on more than windows, maybe the file should be named something other than `win.yml`?

tylerjw commented on moveit_core/collision_detection_bullet/CMakeLists.txt:
```diff
@@ -17,7 +17,7 @@ add_dependencies(${MOVEIT_LIB_NAME} ${catkin_EXPORTED_TARGETS})
 add_library(collision_detector_bullet_plugin src/collision_detector_bullet_plugin_loader.cpp)
 set_target_properties(collision_detector_bullet_plugin PROPERTIES VERSION "${${PROJECT_NAME}_VERSION}")
-target_link_libraries(collision_detector_bullet_plugin ${catkin_LIBRARIES} ${MOVEIT_LIB_NAME})
+target_link_libraries(collision_detector_bullet_plugin ${catkin_LIBRARIES} ${MOVEIT_LIB_NAME} moveit_planning_scene)
```
Shouldn't `moveit_planning_scene` come from `${catkin_LIBRARIES}`?  Why does it need to be manually added here?

Tobias-Fischer replied:
This is the error on macOS if we don't add moveit_planning_scene:
```
 Undefined symbols for architecture x86_64:
  "planning_scene::PlanningScene::setActiveCollisionDetector(std::__1::shared_ptr<collision_detection::CollisionDetectorAllocator> const&, bool)", referenced from:
      collision_detection::CollisionDetectorFCLPluginLoader::initialize(std::__1::shared_ptr<planning_scene::PlanningScene> const&, bool) const in collision_detector_fcl_plugin_loader.cpp.o
ld: symbol(s) not found for architecture x86_64
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
```
(see https://github.com/RoboStack/moveit/runs/2347862944?check_suite_focus=true)

Tobias-Fischer replied:
I'm not sure why it should come from `${catkin_LIBRARIES}`. It's an internal dependency, and my bet is that it just happens that it's not built in the "right" order in some cases. Adding the dependency makes it explicit that `moveit_planning_scene` needs to be built first.

tylerjw commented on moveit_core/collision_detection_fcl/CMakeLists.txt:
```diff
@@ -6,12 +6,12 @@ add_library(${MOVEIT_LIB_NAME}
 set_target_properties(${MOVEIT_LIB_NAME} PROPERTIES VERSION "${${PROJECT_NAME}_VERSION}")
 
-target_link_libraries(${MOVEIT_LIB_NAME} moveit_collision_detection ${catkin_LIBRARIES} ${urdfdom_LIBRARIES} ${urdfdom_headers_LIBRARIES} ${LIBFCL_LIBRARIES} ${Boost_LIBRARIES})
+target_link_libraries(${MOVEIT_LIB_NAME} moveit_collision_detection ${catkin_LIBRARIES} ${urdfdom_LIBRARIES} ${urdfdom_headers_LIBRARIES} ${LIBFCL_LIBRARIES} ${LIBCCD_LIBRARIES} ${Boost_LIBRARIES})
```
Where is the `find_package` for this library?

wolfv replied:
I've removed this link, let's see if it still works.

tylerjw commented on moveit_core/collision_detection_fcl/CMakeLists.txt:
```diff
@@ -6,12 +6,12 @@ add_library(${MOVEIT_LIB_NAME}
 add_library(collision_detector_fcl_plugin src/collision_detector_fcl_plugin_loader.cpp)
 set_target_properties(collision_detector_fcl_plugin PROPERTIES VERSION "${${PROJECT_NAME}_VERSION}")
-target_link_libraries(collision_detector_fcl_plugin ${catkin_LIBRARIES} ${MOVEIT_LIB_NAME})
+target_link_libraries(collision_detector_fcl_plugin ${catkin_LIBRARIES} ${MOVEIT_LIB_NAME} moveit_planning_scene)
```
Once again, why does this need to be manually added, shouldn't it come from `${catkin_LIBRARIES}`?

wolfv replied:
it's done the same in line 9 for `moveit_collision_detection`. this looks like it's a sub-library of a catkin package, right?

tylerjw commented on moveit_core/python/tools/CMakeLists.txt:
```diff
@@ -1,10 +1,12 @@
 set(MOVEIT_LIB_NAME moveit_python_tools)
 
+find_package(Python COMPONENTS Interpreter Development REQUIRED)
+
 add_library(${MOVEIT_LIB_NAME}
   src/pybind_rosmsg_typecasters.cpp
 )
 add_dependencies(${MOVEIT_LIB_NAME} ${catkin_EXPORTED_TARGETS})
-target_link_libraries(${MOVEIT_LIB_NAME} ${catkin_LIBRARIES})
+target_link_libraries(${MOVEIT_LIB_NAME} ${catkin_LIBRARIES} ${PYTHON_LIBRARIES})
```
Does python need to be added to the package.xml and to the dependencies listed in  for this?  Does python need to be added to the dependencies in the `catkin_package` in moveit_core?  Should the find_package be moved into the moveit_core CMakeLists.txt for this?

wolfv replied:
I tried adding python, but cmake still couldn't find it... Do you have an idea how to fix this? For now, I just silence the find_package call and that works fine. 

Linking Python is needed on macOS. Alternatively I think we could add a flag to allow undefined symbols.

tylerjw replied:
You probably still need the `find_package` but you also probably need to add it as a dependency to package.xml so it can be installed by the system in order to find and link.

If you quiet it does that mean it will silently fail when it isn't installed?

Tobias-Fischer replied:
It is now listed in the `package.xml` and thus this comment can be resolved.

tylerjw reviewed (commented):
Some minor comments.  Once this is ready we should run the pre-release test on it also to try to make sure none of these changes will affect a future release.

tylerjw commented:
It looks like there is now a catkin_lint error:

```
moveit_core: CMakeLists.txt: error: unconfigured build_depend on 'pluginlib'
       * You declare a build dependency on another package but neither
       * call find_package() nor have it listed as catkin component in
       * the find_package(catkin) call.
```

Tobias-Fischer commented:
Hmm, I'm not sure what this error is about:
```
<?xml version="1.0" encoding="UTF-8"?>
<testsuites tests="3" failures="1" disabled="0" errors="0" timestamp="2021-04-15T21:19:46" time="22.616" name="AllTests">
  <testsuite name="MoveGroupTestFixture" tests="3" failures="1" disabled="0" errors="0" time="22.616">
    <testcase name="PathConstraintCollisionTest" status="run" time="6.016" classname="MoveGroupTestFixture">
      <failure message="/home/runner/work/moveit/moveit/.work/target_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:255&#x0A;      Expected: planning_scene_interface_.getObjects().size()&#x0A;      Which is: 1&#x0A;To be equal to: std::size_t(0)&#x0A;      Which is: 0&#x0A;Google Test trace:&#x0A;/home/runner/work/moveit/moveit/.work/target_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:155: PathConstraintCollisionTest" type=""><![CDATA[/home/runner/work/moveit/moveit/.work/target_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:255
      Expected: planning_scene_interface_.getObjects().size()
      Which is: 1
To be equal to: std::size_t(0)
      Which is: 0
Google Test trace:
/home/runner/work/moveit/moveit/.work/target_ws/src/moveit/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp:155: PathConstraintCollisionTest]]></failure>
    </testcase>
    <testcase name="CartPathTest" status="run" time="12.1" classname="MoveGroupTestFixture" />
    <testcase name="JointSpaceGoalTest" status="run" time="4.5" classname="MoveGroupTestFixture" />
  </testsuite>
</testsuites>
```

Tobias-Fischer commented:
Cool, seems like all checks are passing now :)

tylerjw commented:
> Cool, seems like all checks are passing now :)

Now I just need to follow up with that small PR I promised.  I'll ping you when that is done.

tylerjw commented:
@v4hn Here is my PR to change the interface to a simple getter: https://github.com/RoboStack/moveit/pull/1  Please review that PR to see if you'd like me to change any part of it before it becomes part of this PR?

wolfv commented:
I tried my best to re-order and squash the commits into a meaningful history. Please let me know what you think! 

I am happy to change it further.

v4hn commented on .github/workflows/cross_platform_ci.yml:
```diff
@@ -0,0 +1,101 @@
+                 -G "Ninja"
+
+        ninja
+        ninja run_tests
```
I'm sure we don't need to include a Linux build in this workflow at all. We have full Linux CI setup for everything we need.

Tobias-Fischer replied:
I would like to see the Linux build based on RoboStack included in the new workflow. It would be much nicer to catch possible issues early on in the new workflow, rather than in our periodic rebuilds in the RoboStack repo. For example, our workflow was able to detect https://github.com/ros-planning/moveit/pull/2473.
Both @wolfv and I would be happy to address any potential issues that arise from the new workflow on Linux build directly before merging in the relevant PR - it would be much less work for us than post-hoc fixes.
However, if you feel strongly about it, at least I'd be happy to remove the Linux build if that increases the chances of getting the PR merged so we can at least test on Win+macOS.

v4hn commented on .github/ci_env.yml:
```diff
@@ -0,0 +1,53 @@
+name: test
```
This file is only necessary for your proposed crossplatform CI check and I believe we usually pull dependencies via `rosdep`.
If this is necessary for your workflow, please give it an an appropriate name `ci_cross_platform_env.yml`.

Tobias-Fischer replied:
We are now using a much more minimal `ci_cross_platform_env.yml` and install the dependencies using rosdep (on both Windows and Unix) as suggested.

v4hn reviewed (commented):
If we actually add this new cross-platform workflow, it should probably remain an optional feature for the foreseeable future.

CI setups tend to break often due to bit rot. Does anyone of you plan to maintain that workflow and provide fixes if it breaks or do you expect the moveit maintainer team to take over? I believe there is currently no maintainer with an intrinsic motivation to fix Windows builds.

Were these questions discussed at a maintainer meeting in the past?

wolfv commented:
Hi @v4hn, I am happy to rename the environment file.
I am also happy to volunteer to keep this CI pipeline running for the foreseeable future, and I think @Tobias-Fischer is also happy to help out.

It's a similar situation to the one we're in where we need to keep patches up-to-date in the robostack repositories, which is quite a bit of work and could be drastically reduced if we could get more people to adopt the packages in the CI infrastructure.

I think supporting (at least CI-wise) macOS and Windows is a great thing, and as you can see the issues related to windows compatibility were not so big after all. Having the CI would mean that it wouldn't accidentally be broken again in the future.

I am happy to join a maintainer meeting if there is one if you want me to chime in.

Also, I am also happy to invite anyone from the MoveIt team to join our little RoboStack get-together (maybe one of those dates work out for you?): https://doodle.com/poll/fi4298qgedknicg3

Tobias-Fischer commented:
I can confirm that I'd be happy to maintain the RoboStack CI. It'd be great to see you guys at our get together!

v4hn commented:
Our next maintainer meeting will probably take place on May, 27th 4pm UTC+2. It would be great if one of you could join and talk about your work here. We will announce the meeting in time [on discourse](https://discourse.ros.org/c/moveit).

Whether this PR will be merged before then also depends on whether some other maintainer working on the CI will get to review this. I'm happy with the actual source changes now, but I don't have the time to properly review/test the GHA changes.
If you could move the proposed CI workflow into a separate pull-request, we could already merge the source changes.

rhaschke commented on moveit_core/CMakeLists.txt:
```diff
@@ -46,15 +42,21 @@ else()
+if (fcl_FOUND)
+  set(LIBFCL_LIBRARIES ${FCL_LIBRARIES})
+  set(LIBFCL_INCLUDE_DIRS ${FCL_INCLUDE_DIRS})
+else()
```
Can you add a comment why the old-style pkgconfig way is still required?

Tobias-Fischer replied:
@wolfv - you played with this; could you please address this comment?

wolfv replied:
Because the dcl from Ubuntu doesn't support FCL CMake properly IIRC.

rhaschke commented on moveit_core/package.xml:
```diff
@@ -24,6 +24,7 @@
   <depend>boost</depend>
   <depend>eigen</depend>
   <depend>bullet</depend>
+  <depend condition="$ROS_DISTRO == noetic">python3</depend>
```
Shouldn't there a python dependency then as well for < Noetic?
```suggestion
  <depend condition="$ROS_DISTRO == noetic">python3</depend>
  <depend condition="$ROS_DISTRO != noetic">python2</depend>
```

rhaschke commented on moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h:
```diff
@@ -44,6 +44,18 @@
 #include <boost/function.hpp>
 #include <string>
 
+// Import/export for windows dll's and visibility for gcc shared libraries.
```
Why do you write this boiler plate manually? There is cmake support for this: https://cmake.org/cmake/help/latest/module/GenerateExportHeader.html

Tobias-Fischer replied:
This particular code is taken from @seanyen's great work within Microsoft's ms-iot. I originally looked at the `GenerateExportHeader` approach in https://github.com/ros-planning/moveit/pull/2328 but didn't realise how big of a task it would be to fix the whole of moveit. So we just follow Microsoft's lead in this. @traversaro - you are the Microsoft+CMake expert, how much work do you think it would be to convert this code to use the cmake generated export headers instead?

wolfv replied:
I think the generate export header macro could also work. Is there a strong preference to use that?

v4hn replied:
By his wording I expect Robert would prefer the cmake-generated content and so would I if it's not even less readable than the manual version.
@rhaschke do you have an example for using this macro that includes support for `ROS_BUILD_SHARED_LIBS`?

wolfv replied:
I just tried to implement the necessary changes. However, I found this in the URDF project where @wjwwood argues against using this CMake specific function: https://github.com/ros/urdf/issues/11#issuecomment-369365046

traversaro replied:
> @traversaro - you are the Microsoft+CMake expert, how much work do you think it would be to convert this code to use the cmake generated export headers instead?

Sorry, I had missed this comment. I don't think it would be particularly difficult. Furthermore, I think that if you switch to `GenerateExportHeader`, there is no need to use `ROS_BUILD_SHARED_LIBS`, as CMake will generate the code appropriately taking in consideration if the library is compiled as static or as shared, unless I am missing something.

rhaschke commented on moveit_core/collision_detection/include/moveit/collision_detection/collision_detector_allocator.h:
```diff
@@ -72,10 +72,7 @@ template <class CollisionEnvType, class CollisionDetectorAllocatorType>
-  {
-    return CollisionDetectorAllocatorType::NAME;
-  }
+  virtual const std::string& getName() const = 0;
```
`getName()` is declared as abstract in the base class `CollisionDetectorAllocator` already. Why repeating this here?

tylerjw replied:
This is my fault.  I put this here to make it clear that this is also now a abstract class.  It would make no difference to delete it.

Tobias-Fischer replied:
We are happy either way; please let us know what your preference is.

rhaschke commented on .github/workflows/cross_platform_ci.yml:
```diff
@@ -0,0 +1,101 @@
+        os: [ubuntu-latest, macos-latest, windows-latest]
+    steps:
+    - uses: actions/checkout@v2
+    - name: Build Moveit Core on Unix
```
Can we employ ccache here as well? See our [ci.yaml](https://github.com/ros-planning/moveit/blob/f62c30f61e539541c7a7d0387db75c559b0974e2/.github/workflows/ci.yaml#L75-L82).

Tobias-Fischer replied:
Unfortunately using `ccache` alongside `conda` is still an open issue / experimental; see https://github.com/conda-forge/conda-forge.github.io/issues/389 and https://github.com/conda-forge/compilers-feedstock/issues/31. I'd rather not include this for now, as it might cause unforeseen issues. What do you think @wolfv?

wolfv replied:
yes, I agree. We do have a ccache package on conda-forge and I want to figure out how to make use of it but I would defer this for later. Also this CI pipeline is quite fast as it only builds moveit-core for now. Once we start adding the other moveit packages it would make sense to re-visit ccache.

rhaschke reviewed (changes requested):
Overall this looks good. See remarks.

rhaschke commented on .github/workflows/cross_platform_ci.yml:
```diff
@@ -0,0 +1,101 @@
+
+jobs:
+  run_moveit_core_compilation:
+    runs-on: ${{ matrix.os }}
```
```suggestion
    runs-on: ${{ matrix.os }}
    name: ${{ matrix.os }}
```

Tobias-Fischer replied:
Fixed in https://github.com/ros-planning/moveit/pull/2604/commits/839b4c0c569eb760efe4b5b2a8ffa856fff2c1c5

v4hn commented:
> Our next maintainer meeting will probably take place on May, 27th 4pm [it's 5pm!] UTC+2. It would be great if one of you could join and talk about your work here. We will announce the meeting in time on discourse.

Here's the announcement post for the meeting next week:
https://discourse.ros.org/t/moveit-manipulation-wg-meeting-may-27th/20431
It's going to take place 5pm UTC+2, sorry about the wrong statement before.
It would be good if either of you could attend and present your integration efforts and the purpose of this new CI workflow.

Also, you did not resolve @rhaschke's comments yet.

Tobias-Fischer commented on .github/workflows/cross_platform_ci.yml:
```diff
@@ -0,0 +1,101 @@
+name: Build Moveit Core on all platforms
```
Rosdep and actions name are both done.

Tobias-Fischer commented on .github/workflows/cross_platform_ci.yml:
```diff
@@ -0,0 +1,101 @@
+      run: |
+        export CTEST_OUTPUT_ON_FAILURE=1
+        cd moveit_core
+        ls -al
```
This comment was related to a previous commit and has been fixed since.

Tobias-Fischer commented on .github/workflows/cross_platform_ci.yml:
```diff
@@ -0,0 +1,101 @@
+        ninja
+        ninja run_tests
+
+    - name: Build Moveit Core on Windows
```
This comment was related to a previous commit and has been fixed since.

Tobias-Fischer commented on .github/workflows/cross_platform_ci.yml:
```diff
@@ -0,0 +1,101 @@
+      shell: bash -l -eo pipefail {0}
+      run: |
+        export CTEST_OUTPUT_ON_FAILURE=1
+        cd moveit_core
+
+        mkdir build
+        cd build
```
May I suggest to defer this to a later PR? There are quite a few more patches in the moveit code for the other packages (see https://github.com/RoboStack/ros-noetic/tree/master/patch) to get them to run on Win+macOS, and including all of them in a single PR would make it very hard to review. Our aim is definitely to include all packages though and we'd follow up with more PRs.

Tobias-Fischer commented:
Hi @v4hn, many thanks for your reminder and letting us know about the meeting! 

I addressed a few more comments by @rhaschke, and left comments regarding the others. Unfortunately the meeting is at 1am Brisbane time, so I won't be able to make it. I am not sure whether @wolfv will be able to attend; I'll send him another reminder re this.

wolfv commented:
I'll be there on the 27th. Looking forward to seeing you all.

Tobias-Fischer commented:
Hi @v4hn - are there any more changes you'd like us to do? I think we addressed all of the comments, and @wolfv said that you were positive in the maintainer meeting about the PR.

wolfv commented:
Here is a success story of MoveIt on Windows through RoboStack: https://github.com/compas-dev/compas_fab/issues/85#issuecomment-809443514

Would be great to get this merged :)

v4hn commented on .github/workflows/cross_platform_ci.yml:
```diff
@@ -0,0 +1,106 @@
+    branches:
+      - master
+      - "[kmn]*-devel"
+      - tests_with_conda_pkgs
```
```suggestion
```

v4hn reviewed (commented):
Ok, this is almost ready to merge. Sorry for the many iterations, it's great you are investing so much time here!

Aside from (1) the currently failing tests, (2) the additional branch name that will not be relevant upstream and (3) the manual boiler plate vs cmake-generator aspect this looks good to me.

As we accumulate more and more CI jobs, I am beginning to wonder how the jobs are sorted in the PR overview at the bottom.
[This](https://github.com/github/hub/pull/2023) seems to imply there was a clear order involved, but what I see here right now looks entirely arbitrary:

![foo](https://user-images.githubusercontent.com/680358/123526308-51958c00-d6d7-11eb-875d-7cd3cd64a019.png)

Of course it would be nice to keep the entries grouped by workflow, but maybe I'm missing something?

wolfv commented:
Looks like we should also build moveit_msgs in the tests. I'll take care of that.

wolfv commented:
The division by zero throws an error on MSVC: https://github.com/RoboStack/moveit/runs/2924134601?check_suite_focus=true#step:7:1006

wolfv commented:
Tests are green again: https://github.com/RoboStack/moveit/runs/2924301432?check_suite_focus=true

I removed the branch name.

For the automically generated export helper block -- I think we can do it. Will have a stab at it later or tomorrow.

v4hn commented:
> However, I found this in the URDF project where @wjwwood argues against using this CMake specific function

I know Robert and William agreed to disagree on this in some thread before. :-)

I really don't care about William's second argument on depending on cmake, but do see the point about problematic auto-completion if you don't want to build the project before browsing the code. With MoveIt that can actually take valuable time.
We could probably work around it by providing an empty definition and `__has_include` magic (from c++17...), but it might not be worth the trouble.
At least I would ask you to use the files as generated by cmake and add the corresponding cmake lines as comments.
I'm open to commit the generated headers then.

traversaro commented:
> I'm open to commit the generated headers then.

I am afraid this will not work as intended. `GenerateExportHeader` does the platform detection and the static/shared library detection logic at CMake configuration time, and the header that is generated by it is only valid for that specific platform and with that specific library type (see https://gitlab.kitware.com/cmake/cmake/-/blob/v3.20.5/Modules/exportheader.cmake.in and https://gitlab.kitware.com/cmake/cmake/-/blob/v3.20.5/Modules/GenerateExportHeader.cmake#L316). This is different from the approach done in manually written visibility headers, in which the platform detection is done at C preprocessor time.

traversaro commented:
A few pointers on past discussions on manually writing visibility headers vs using `GenerateExportHeader`  (not sure if it includs the discussion mentioned by @v4hn, probably not): 
* https://github.com/ms-iot/ROSOnWindows/issues/323
* https://github.com/ms-iot/ROSOnWindows/issues/33 
* https://github.com/ros/console_bridge/pull/40#issuecomment-269460721
* https://github.com/ros2/rclcpp/pull/638#discussion_r260472396

TL;DR: As mentioned by @v4hn manully written headers and `GenerateExportHeader`-generated headers have their own relative set of advantages and disadvantages, so it is usually up to the maintainers to chose which option they prefer.

v4hn commented:
Ok, in that case I support Robert's request for autogenerated files.
If possible, please implement this so we can finally merge this request.

traversaro commented:
> 
> Ok, in that case I support Robert's request for autogenerated files. If possible, please implement this so we can finally merge this request.

Perfect, thanks! @wolfv let me know if you want to proceed with your effort that mentioned earlier in the PR or you like me to add `generate_export_header` support.

wolfv commented:
@traversaro if you could help me out here it would be much appreciated :) My current efforts look like this: https://github.com/RoboStack/moveit/compare/tests_with_conda_pkgs...RoboStack:gen_headers?expand=1

But didn't have the time to test it properly yet (and looks like CI fails).

traversaro commented:
> 
> 
> @traversaro if you could help me out here it would be much appreciated :) My current efforts look like this: https://github.com/RoboStack/moveit/compare/tests_with_conda_pkgs...RoboStack:gen_headers?expand=1
> 
> But didn't have the time to test it properly yet (and looks like CI fails).

The fixes for that branch to work correctly are provided in https://github.com/RoboStack/moveit/pull/3 .

v4hn reviewed (approved):
Assuming CI passes (I'll not wait for the last three jobs anymore today), this looks good to merge.
Please rebase the commits to remove the fixups and the auxiliary branch name in the CI config.

I would prefer to have this merged via merge commit as a number of semi-independent changes are also fixed here.

wolfv commented:
I've just rebased, removed the very last commit (that re-activated building on our branch) and attributed the CMake changes correctly to @traversaro.

Would be awesome to see this merged. Thanks for all the helpful reviews!

v4hn commented:
Thank you for the contributions and your promise to take care of the new CI job in the future (and/or expand it to more than moveit_core).
This has been lying around for long enough, so I'll merge it now to avoid other PRs breaking it again.

wolfv commented:
thanks for merging!
